spring-session/src/main/java/org/springframework/session/SafeRetrievingSessionRepository.java

```diff
@@ -0,0 +1,185 @@
+ *
+ * @param <S> the type of {@link Session} being managed
+ * @author Vedran Pavic
+ * @since 1.3.0
+ */
+public class SafeRetrievingSessionRepository<S extends Session>
+		implements FindByIndexNameSessionRepository<S> {
+
+	private static final Log logger = LogFactory.getLog(SafeRetrievingSessionRepository.class);
+
+	private final FindByIndexNameSessionRepository<S> delegate;
+
+	private final Set<Class<? extends RuntimeException>> ignoredExceptions;
+
+	private boolean deleteOnIgnoredException = true;
+
+	private boolean logIgnoredException;
+
+	/**
+	 * Create a new {@link SafeRetrievingSessionRepository} instance backed by a
+	 * {@link FindByIndexNameSessionRepository} delegate.
+	 * @param delegate the {@link FindByIndexNameSessionRepository} delegate
+	 * @param ignoredExceptions the set of exceptions to ignore
+	 */
+	public SafeRetrievingSessionRepository(
+			FindByIndexNameSessionRepository<S> delegate,
+			Set<Class<? extends RuntimeException>> ignoredExceptions) {
+		Assert.notNull(delegate, "Delegate must not be null");
+		Assert.notEmpty(ignoredExceptions, "Ignored exceptions must not be empty");
+		this.delegate = delegate;
+		this.ignoredExceptions = ignoredExceptions;
+	}
+
+	/**
+	 * Create a new {@link SafeRetrievingSessionRepository} instance backed by a
+	 * {@link SessionRepository} delegate.
+	 * @param delegate the {@link SessionRepository} delegate
+	 * @param ignoredExceptions the set of exceptions to ignore
+	 */
+	public SafeRetrievingSessionRepository(SessionRepository<S> delegate,
+			Set<Class<? extends RuntimeException>> ignoredExceptions) {
+		this(new FindByIndexNameSessionRepositoryAdapter<S>(delegate), ignoredExceptions);
+	}
+
+	/**
+	 * Set whether session should be deleted after ignored exception has occurred during
+	 * retrieval.
+	 * @param deleteOnIgnoredException the flag to indicate whether session should be
+	 * deleted
+	 */
+	public void setDeleteOnIgnoredException(boolean deleteOnIgnoredException) {
+		this.deleteOnIgnoredException = deleteOnIgnoredException;
+	}
+
+	/**
+	 * Set whether ignored exception should be logged.
+	 * @param logIgnoredException the flag to indicate whether to log ignored exceptions
+	 */
+	public void setLogIgnoredException(boolean logIgnoredException) {
+		this.logIgnoredException = logIgnoredException;
+	}
+
+	public S createSession() {
+		return this.delegate.createSession();
+	}
+
+	public void save(S session) {
+		this.delegate.save(session);
+	}
+
+	public S getSession(String id) {
+		try {
+			return this.delegate.getSession(id);
+		}
+		catch (RuntimeException e) {
+			if (isIgnoredException(e)) {
+				if (this.logIgnoredException) {
+					logger.warn("Error occurred while retrieving session " + id + ": "
+							+ e);
+				}
+				if (this.deleteOnIgnoredException) {
+					this.delegate.delete(id);
+				}
+				return null;
+			}
+			throw e;
+		}
+	}
+
+	public void delete(String id) {
```

Comment:
Thanks for giving this a test run @davidmelia. It's a problem with this code since it doesn't take into account that delegate's `delete` operation (or any other for that matter) could internally call an operation that can result in a deserialization error.

I guess this is a case in point for implementing some sort of deserialization strategy vs the session repository decorator pattern.